packages/trader/src/App/Containers/SetResidenceModal/set-residence-modal.jsx

```diff
@@ -0,0 +1,147 @@
+import classNames       from 'classnames';
+import {
+    Button,
+    Dialog }            from 'deriv-components';
+import {
+    Formik,
+    Form }              from 'formik';
+import PropTypes        from 'prop-types';
+import React            from 'react';
+import { localize }     from 'App/i18n';
+import { connect }      from 'Stores/connect';
+import { website_name } from 'App/Constants/app-config';
+import ResidenceForm    from './set-residence-form.jsx';
+import                       'Sass/app/modules/set-residence.scss';
+
+const residenceInitialValues = { residence: '' };
+
+const validateResidence = (values, residence_list) => {
+    const errors = {};
+
+    if (!values.residence) {
+        errors.residence = true;
+    } else {
+        const index_of_selection = residence_list.findIndex(item => (
+            item.text.toLowerCase() === values.residence.toLowerCase()
+        ));
+
+        if (index_of_selection === -1 || residence_list[index_of_selection].disabled === 'DISABLED') {
+            errors.residence = localize('Unfortunately, {{website_name}} is not available in your country.', { website_name });
+        }
+    }
+
+    return errors;
+};
+
+class SetResidence extends React.Component {
+    onSetResidenceComplete = (error) => {
+        // TODO: Proper error handling (currently we have no place to put the message)
+        if (error) {
+            throw Error(error);
+        }
+        // Handle lower level modal controls due to overriding modal rendering
+        this.props.isModalVisible(false);
+        this.props.enableApp();
+    };
+
+    render() {
+        const { onSetResidence, residence_list } = this.props;
+
+        const validateResidencePassthrough = (values) => validateResidence(values, residence_list);
+        const onSetResidencePassthrough = (values) => {
+            const index_of_selection = residence_list.findIndex(item => (
```

Comment:
`.find()`? we could have a shared function for getting residence since we are repeating this logic in a lot of places